### PR TITLE
Qualcomm AI Engine Direct - support embedding op

### PR DESCRIPTION
**Pull request by haowhsu-quic**
summary:
- support embedding op with int32 index input
- make mobilebert / llama2 be fully delegated
- add requantize passes for mixed precision
- bug fixes

### CODE REVIEW

File: backends/qualcomm/tests/test_qnn_delegate.py [hunk 1]
```diff
@@ -1307,8 +1302,9 @@ def test_mobilebert(self):
             msg = json.loads(conn.recv())
             cpu, htp = msg["CPU"], msg["HTP"]
             for k, v in cpu.items():
-                self.assertLessEqual(abs(v[0] - htp[k][0]), 1)
+                self.assertLessEqual(abs(v[0] - htp[k][0]), 2)
 
+    @unittest.skip("failure caused by convert_pt2e")
```
Comment (jerryzh168) [inline backends/qualcomm/tests/test_qnn_delegate.py:1307]: what is the failure?

**Review by cccclai — APPROVED** [review]

**Comment by cccclai** [comment]
Two main comments:
- Is there a way to repro the error on our side? maybe it's an edge case we'd need to fix
- Can we leave a todo (maybe an expected failing unit test)?

**Comment by haowhsu-quic** [comment]
> Two main comments:
> 
> * Is there a way to repro the error on our side? maybe it's an edge case we'd need to fix
> * Can we leave a todo (maybe an expected failing unit test)?

Thanks for reviewing! I've added TODO items for next action.

It could be reproduced with following patch:
```diff
[examples/qualcomm/scripts/mobilebert_fine_tune.py]
@@ -58,10 +58,18 @@ def accuracy_per_class(preds, goldens, labels):
def get_dataset(data_val):
    # prepare input data
    inputs, input_list = [], ""
-   # max_position_embeddings defaults to 512
-   position_ids = torch.arange(512).expand((1, -1)).to(torch.int32)
    for index, data in enumerate(data_val):
        data = [d.to(torch.int32) for d in data]
+       # input_ids, attention_mask, token_type_ids
+       inputs.append((*data[:2], torch.zeros(data[0].size(), dtype=torch.int32)))
-       # input_ids, attention_mask, token_type_ids, position_ids
        inputs.append(
            (
                *data[:2],
                torch.zeros(data[0].size(), dtype=torch.int32),
-               position_ids[:, : data[0].shape[1]],
            )
        )
        input_text = " ".join(
            [f"input_{index}_{i}.raw" for i in range(len(inputs[-1]))]
        )
@@ -204,9 +212,6 @@ def get_fine_tuned_mobilebert(artifacts_dir, pretrained_weight, batch_size):
            map_location=torch.device("cpu"),
        ),
    )
+   # hack for changing dtype of "position_ids" from int64 to int32
+   sub_module = model.mobilebert.embeddings
+   sub_module.position_ids = sub_module.position_ids.to(torch.int32)

    return model.eval(), dataloader_val, labels
```

**Comment by cccclai** [comment]
Thanks for the update. Seems like some llama related CI jobs start failign. Those changes look legit but Is it okay to remove changes in examples/models/llama2/model.py to get CI green? We can do a seperate pr for this.

**Comment by haowhsu-quic** [comment]
Thank you, I moved the datatype casting from `examples/models/llama2/model.py` to our own script as mobilebert does.

**Pull request closed without merging**